x/txfees/keeper/feedecorator_test.go

```diff
@@ -157,23 +174,60 @@ func (suite *KeeperTestSuite) TestFeeDecorator() {
 			gasRequested: mempoolFeeOpts.HighGasTxThreshold,
 			isCheckTx:    true,
 			expectPass:   true,
+			baseDenomGas: false,
 		},
 	}
 
 	for _, tc := range tests {
-
 		suite.ctx = suite.ctx.WithIsCheckTx(tc.isCheckTx)
 		suite.ctx = suite.ctx.WithMinGasPrices(tc.minGasPrices)
 
-		tx := legacytx.NewStdTx([]sdk.Msg{}, legacytx.NewStdFee(
-			tc.gasRequested,
-			tc.txFee,
-		), []legacytx.StdSignature{}, "")
+		// TxBuilder components reset for every test case
+		txBuilder := suite.clientCtx.TxConfig.NewTxBuilder()
+		priv0, _, addr0 := testdata.KeyTestPubAddr()
+		acc1 := suite.app.AccountKeeper.NewAccountWithAddress(suite.ctx, addr0)
+		suite.app.AccountKeeper.SetAccount(suite.ctx, acc1)
+		msgs := []sdk.Msg{testdata.NewTestMsg(addr0)}
+		privs, accNums, accSeqs := []cryptotypes.PrivKey{priv0}, []uint64{0}, []uint64{0}
+		signerData := authsigning.SignerData{
+			ChainID: suite.ctx.ChainID(), 
+			AccountNumber: accNums[0], 
+			Sequence: accSeqs[0]}
+
+		gasLimit := tc.gasRequested
+		sigV2, _ := clienttx.SignWithPrivKey(
+			1, 
+			signerData,
+			txBuilder, 
+			privs[0], 
+			suite.clientCtx.TxConfig,
+			accSeqs[0])
+
+		simapp.FundAccount(suite.app.BankKeeper, suite.ctx, addr0, tc.txFee)
+		
+		txBuilder.SetMsgs(msgs[0])
+		txBuilder.SetSignatures(sigV2)
+		txBuilder.SetMemo("")
+		txBuilder.SetFeeAmount(tc.txFee)
+		txBuilder.SetGasLimit(gasLimit)
+
+		tx := txBuilder.GetTx()
 
 		mfd := keeper.NewMempoolFeeDecorator(*suite.app.TxFeesKeeper, mempoolFeeOpts)
-		antehandler := sdk.ChainAnteDecorators(mfd)
-		_, err := antehandler(suite.ctx, tx, false)
+		dfd := keeper.NewDeductFeeDecorator(*suite.app.TxFeesKeeper, *suite.app.AccountKeeper, *suite.app.BankKeeper, *suite.app.FeeGrantKeeper)
+		antehandlerMFD := sdk.ChainAnteDecorators(mfd, dfd)
+		_, err := antehandlerMFD(suite.ctx, tx, false)
+
 		if tc.expectPass {
+			if tc.baseDenomGas && !tc.txFee.IsZero() {
+				moduleAddr := suite.app.AccountKeeper.GetModuleAddress(types.FeeCollectorName)
+				suite.Require().Equal(tc.txFee[0], suite.app.BankKeeper.GetBalance(suite.ctx, moduleAddr, baseDenom), tc.name)
+				suite.app.BankKeeper.SendCoinsFromModuleToAccount(suite.ctx, types.FeeCollectorName, addr0, tc.txFee)
```

Comment:
why are we sending funds out of the module to an account in the tests?

Reply:
Ah that makes, yeah there is a better way!

At the beginning of every test cases processing code, do:
```
suite.SetupTest(false)
suite.ExecuteUpgradeFeeTokenProposal(uion, uionPoolId)
```

And then every test will start on a brand new, clean state

Reply:
Awesome, adding in next commit!